### ISSUE TITLE
Simplify output methods

### ISSUE THREAD

**Issue by mahrud** [Macaulay2/M2#1529]
Currently we have the following output methods:
```m2
net, info, html, htmlWithTex, markdown, mathML, tex, texMath
```
There are also some internal ones that are not exported, like `texWrap`.

Now that math in the web documentation is rendered with KaTeX, I think we can simplify some of these. For instance, `html` applied to a non-hypertext object should simply call `tex` on the object and wrap it in appropriate delimiters (`$`/`$$`, or maybe `\(`/`\[`). This is basically what `htmlWithTex` does, so I think we can basically merge them by fold in the before/after printing features of `htmlWithTex` into `html`. In other words, all HTML is HTML with tex now, so there's no distinction.

Is there a reason not to do this?

cc: @pzinn

**Comment by pzinn** [comment]
besides my answer in another thread:
> we definitely should discuss this in more detail at some point, but my short answer: I think html should indeed produce html 
> code, not tex code. if one wants mixed output, a method like htmlWithTex would be appropriate, yes. or the user should be 
> allowed to directly insert tex code into his HTML using TEX or something like that, but it's a different story. I didn't follow your
> recent changes, so am not sure if they contradict the philosophy above.

I want to say one word about `texWrap`: right now the main use of `htmlWithTex` is for my web app. unfortunately the web app cannot use standard delimiters $ for tex math, it uses instead its own tags (ascii<32). That's why in my original version of the code I had made the delimiters used by `tex` mutable objects, so `htmlWithTex` just changed these before calling `tex`. Dan protested (rightly) that this made the code not reentrant or thread-safe or what not, so instead of really fixing the problem I temporarily introduced `texWrap` which is... the same as `tex` except with modifyable delimiters. This code will go, eventually, but before that I want to clarify some of the other related issues, such as the `html` issue that started this conversation.

**Comment by pzinn** [comment]
before I forget, there's also one side issue which is that `htmlWithTex String` and `html String` differ, as well as `htmlWithTex Net` and `html Net`.

**Comment by mahrud** [comment]
> before I forget, there's also one side issue which is that `htmlWithTex String` and `html String` differ, as well as `htmlWithTex Net` and `html Net`.

Yeah, I was hoping to talk about those. I think htmlWithTex does the right thing for Net, but I'm not sure about String. In particular, why does it put it in <pre>?

Descent is also something that I think htmlWithTex does right, and I'm wondering why the code isn't just implemented for html.

Also, are the webapp tags only used so the webapp can tell where to split the M2 output?

**Comment by pzinn** [comment]
no, the tags are for more than that. most importantly, they're there to distinguish ordinary text output (which M2 still produces in various uncontrolled circumstances, such as error messages) and HTML output.
one radical change I could make is give up on TeX tags entirely and go back to using just $. that would mean removing `texWrap`. But then I need to be absolutely sure that $ doesn't get used in HTML mode, which means `html` must encode those.
this is probably a good idea anyway if the use of KaTeX is generalized. is this the case currently? (can't check right now)

**Comment by pzinn** [comment]
anyway I'm ready to concede and let `html` produce TeX inside of HTML, since AFAICT the decision has already been made. But if I'm to merge `html` and `htmlWithTex`, I'm worried about types for which it's not clear which output is better (in fact, there might even be cases where both outputs are valid, and we're depriving the user of the choice. but I digress). Simple example, the type "Type": clearly the output is just a string. so should it also have HTML output, rather than TeX output?

**Comment by pzinn** [comment]
> Yeah, I was hoping to talk about those. I think htmlWithTex does the right thing for Net, but I'm not sure about String. In particular, why does it put it in `<pre>`?
because oherwise white space is not preserved. if the string is "H         I", then the web interface should print it as is, with all the spaces. there's something ambiguous about the treatment of strings in M2 `html`: they're also used internally by `Hypertext` elements as the content of the HTML element -- which, you would argue, shouldn't preserve white space.

**Comment by pzinn** [comment]
also, does `html` encode HTML special characters (e.g. <>) of its KaTeX content? because `htmlWithTex` doesn't.

**Comment by pzinn** [comment]
answering myself...
one of the problem is the new `${prefix}` in URLs... If I latex KaTeX use `$` as a delimiter, it gets really confused because of that...

**Comment by pzinn** [comment]
does this even work? I can't test easily, but do links in examples work even in normal mode?
e.g.,
`help ofClass`

**Comment by DanGrayson** [comment]
How would I observe a new URL with a dollar sign in it?

**Comment by mahrud** [comment]
I think Paul's point is that `$` should always be escaped in html output, unless it is a LaTeX delimiter. Maybe we could say that only `\( .. \)` and `\[ .. \]` can be used to avoid this, and escape all `$`. I'll think more about it.

**Comment by pzinn** [comment]
right. my suspicion is, if one recreates the current documentation, the page for `ofClass` will be messed up because the example contains this `${prefix}`. 

that being said, if I have to revert to normal TeX delimiters, I'd rather stick to `$` rather than any of the alternatives.

**Comment by pzinn** [comment]
slightly off-topic, but continuing the discussion of $ in documentation:
```
help "replace(String,String,String)"
```
has a lonely $ sign which KaTeX won't like...

**Comment by mahrud** [comment]
> slightly off-topic, but continuing the discussion of $ in documentation:
> 
> ```
> help "replace(String,String,String)"
> help get
> ```
> 
> have lonely $ signs which KaTeX won't like...

It looks fine:
![image](https://user-images.githubusercontent.com/3147718/96629255-fc4be400-12d8-11eb-98cb-5d530927e7d8.png)
KaTeX doesn't touch anything inside `<tt>`, `<pre>`, and a few other tags.

**Comment by pzinn** [comment]
yeah I noticed a couple of those as well.
anyway, I think this destroys my last hope of using $ in the web interface. there's no way the client can keep track of when we are inside or not of a <tt> <pre> or what not (it's not going to start parsing the html...)
so I may have to rethink my merge of `html` and `htmlWithTex`.

**Comment by mahrud** [comment]
I don't understand the problem, because KaTeX is not confused about random $s in code, and your client can take an object of type TEX or Expression and feed it directly to the KaTeX api instead of wrapping it in $.

**Comment by pzinn** [comment]
what do you mean by "client"? I meant the web client (i.e., the JS code). I'm not quite sure what you suggest.
note that for the documentatoin you're using KaTeX on a static page. The html in the web interface gets fed dynamically (typically, chopped up in successive packets). it's quite different.

**Comment by pzinn** [comment]
one solution would be to replace $ -> &dollar; inside these tags `<tt> <pre>` etc. would make my life simpler.
anyway please take a look at [merge/htmlwithtex](https://github.com/pzinn/M2/tree/merge/htmlwithtex) (in particular, `methods html`) and let me know if this is what you had in mind.

**Comment by pzinn** [comment]
in the absence of any comments, I will soon test this branch on the web server, and if all goes well, submit a PR.

**Comment by mahrud** [comment]
You can also url encode everything except for the parts that need to be processed by KaTeX on the server side of the web interface, then after KaTeX is parsed, url decode and display them normally.

**Comment by pzinn** [comment]
at the moment, the server side of the interface does not do anything except transmit back and forth between the M2 process and the client. I'd rather keep it this way. In any case, I think I've got the web interface under control now.
In the process of merging `htmlWithTex` and `html`, I have however changed the behavior of `html` quite a bit. That's the part that affects M2 even outside the web interface -- hopefully, only in a good way, along  the lines that you suggested.

**Comment by pzinn** [comment]
> 
> This is an issue, though, but it is unrelated:
> ![image](https://user-images.githubusercontent.com/3147718/96629544-59e03080-12d9-11eb-91b4-a2c2172304b3.png)
> ![image](https://user-images.githubusercontent.com/3147718/96629564-5f3d7b00-12d9-11eb-81fe-6c8888a98efc.png)

should we fix that issue before the next release? everything else looks good to me.

**Comment by DanGrayson** [comment]
Yes, let's fix that.  Is it a problem also in previous versions?

**Comment by pzinn** [comment]
no I think it's due to recent documentation changes. found a couple more bugs, preparing a PR now.

**Comment by pzinn** [comment]
Pretty much all the issues in this thread have been resolved.
Though I still think the class `TEX` should be retired.
A related issue is the following: right now there's some new code (which I'm assuming you wrote @mahrud ?)  which converts automatically a bunch of TeX commands like `\bf` to their HTML analogues like `<b>`. I don't think this should happen inside `$..$`.
For example the tutorial in the documentation
```
help "Tutorial: Canonical Embeddings of Plane Curves and Gonality"
```
uses `${\bf P}^1$` which gets incorrectly converted to `$<b>P</b>^1$`. At least in my web interface the result is incorrect formatting.

**Comment by mahrud** [comment]
Sure, but probably it's easier to just change that to `{\mathb P}`, or better yet, `\PP` instead.

**Comment by pzinn** [comment]
this may be true in this particular example, but the fact remains that people should be allowed to use `\bf` in their TeX code without it being converted to `<b>`.

**Comment by mahrud** [comment]
I disagree. Using `\bf` is frowned upon even in normal situations. They should use either `\mathbf` or `\textbf`. The conversion for `\bf, \it, \tt, \em` is only there for compatibility with some old documentation nodes.

**Comment by pzinn** [comment]
with the same logic we should prevent the use of `$` and `$$`. Originally, AFAICT, `tex` was meant for plain TeX, not LaTeX.
overall, I think that anything that is accepted by KaTeX should be left as is.

**Comment by pzinn** [comment]
In fact, let's look at the code generated by this conversion:
```
i1 : Z=help "Tutorial: Canonical Embeddings of Plane Curves and Gonality";

i2 : peek(Z#1#2)

o2 = PARA{TEX{"The gonality of a curve is defined to be the smallest degree of a morphism from the curve to the projective line $", BOLD{"P"}, "^1$. It is known that a curve $C$
     ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
     of genus $g$ admits a map to $", BOLD{"P"}, "^1$ of degree at most $[(g+3)/2]$.  Further, if $C$ is $d$-gonal, then in its canonical embedding $C$ lies on a rational normal
     ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
     scroll of dimension $d-1$, and the free resolution of the homogeneous coordinate ring of the scroll is a subcomplex of the free resolution of the homogeneous coordinate ring
     ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
     of  $C$.  Thus for example the $2$-linear part of that resolution has length at least $g-d$, and ``Green's Conjecture'' states that if one computes Clifford index instead of
     ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
     gonality, a slight refinement, then equality holds. For example, Green's conjecture predicts that the resolution of the homogeneous coordinate ring of a general curve of
     ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
     genus 7 and gonality 4 is:"}}

```
you can see how dangerous this is: the text node got cut in half by the `BOLD{"P"}`, which will thoroughly confuse KaTeX.

**Comment by mahrud** [comment]
I think this issue is mostly done, right?

The other things I think would be nice are:
- [ ] split expressions.m2 into smaller pieces:
  - one file for all objects that inherit from Holder or Expression (what's the difference?)
  - one file for AfterPrints
  - find a way to define reduce the number of `texMath` methods by defining `texMath Expression`, then defining `texMath Thing := x -> texMath expression x`
  - generally clean it up, remove commented code, etc.
- [ ] clean up latex.m2
- [ ] clean up mathml.m2
- [ ] similar to the item above about texMath, find a way to define a single printing method for each type, then have tex, texMath, net, info, etc. have a standard way of taking the output of that method and returning the appropriate type of output

**Comment by pzinn** [comment]
yeah I guess this issue can be closed.

concerning the last point, isn't that what expressions are about?

**Comment by mahrud** [comment]
> concerning the last point, isn't that what expressions are about?

Yes, but I guess I don't know why there are 82 methods for `texMath` and 90 for `net` but only 38 for `expression`. Seems like some stuff might be missing.

**Comment by pzinn** [comment]
As a wise man once said:
There is now a consistent logic for expression methods, with a proper Expression following a simple tree structure. Types fall into two categories:
- Types whose expression is a Holder of themselves (expression is "turned off"); then these types must have output routines toString/net/texMath/html/etc. These objects are considered "elementary" -- they cannot be broken down further; they're the leaves of the tree. ex:
code(expression,BettiTally)
- Types which have an actual expression (and no output routines). These types are the nodes of the tree; typically, their expression acts recursively. ex:
code(expression,MutableMatrix)

it's normal that there would be more nodes than leaves.

**Comment by pzinn** [comment]
If you really wanted say `texMath` to have methods for `Expression` types only (and for all the rest to be `texMath @@ expression`) then either you'd have to create a dummy `Expression` type for every type that currently has `texMath := hold`, or to have a very long list of cases for `texMath Holder`. Neither solution is satisfactory. But maybe you have an even more radical idea in mind?

**Comment by mahrud** [comment]
This is not a complete reply, but can the commented lines involving Holder2 be deleted? They are all over the place.
